cpp/src/kmeans/common.cuh

```diff
@@ -242,61 +312,126 @@ void minClusterAndDistance(
     // input dataset
     auto datasetView = X.template view<2>({ns, n_features}, {dIdx, 0});
 
-    // distanceView [ns x n_clusters]
-    auto distanceView =
-      pairwiseDistance.template view<2>({ns, n_clusters}, {0, 0});
-
     // minClusterAndDistanceView [ns x n_clusters]
     auto minClusterAndDistanceView =
       minClusterAndDistance.template view<1>({ns}, {dIdx});
 
-    // calculate pairwise distance between cluster centroids and current batch
-    // of input dataset
-    kmeans::detail::pairwiseDistance(handle, datasetView, centroids,
-                                     distanceView, workspace, metric, stream);
-
-    // argmin reduction returning <index, value> pair
-    // calculates the closest centroid and the distance to the closent centroid
-    cub::KeyValuePair<IndexT, DataT> initial_value(
-      0, std::numeric_limits<DataT>::max());
-    MLCommon::LinAlg::coalescedReduction(
-      minClusterAndDistanceView.data(), distanceView.data(),
-      distanceView.getSize(1), distanceView.getSize(0), initial_value, stream,
-      false,
-      [=] __device__(const DataT val, const IndexT i) {
-        cub::KeyValuePair<IndexT, DataT> pair;
-        pair.key = i;
-        pair.value = val;
-        return pair;
-      },
-      [=] __device__(cub::KeyValuePair<IndexT, DataT> a,
-                     cub::KeyValuePair<IndexT, DataT> b) {
-        return (b.value < a.value) ? b : a;
-      },
-      [=] __device__(cub::KeyValuePair<IndexT, DataT> pair) { return pair; });
+    auto L2NormXView = L2NormX.template view<1>({ns}, {dIdx});
+
+    // tile over the centroids
+    for (auto cIdx = 0; cIdx < n_clusters; cIdx += centroidsBatchSize) {
+      // # of centroids for the current batch
+      auto nc = std::min(centroidsBatchSize, n_clusters - cIdx);
+
+      // centroidsView [nc x n_features] - view representing the current batch
+      // of centroids
+      auto centroidsView =
+        centroids.template view<2>({nc, n_features}, {cIdx, 0});
+
+      if (metric == MLCommon::Distance::EucExpandedL2) {
```

Comment:
This sucks. I see the reason for duplication due to templating of 'sqrt' parameter in the fused-l2-nn prim. Let me make it a runtime option so as to reduce this code duplication.

Reply:
Done